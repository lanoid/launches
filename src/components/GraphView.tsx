import * as React from 'react';
import launchReducer from '../helpers/launchReducer';
import Chart = require('chart.js');

interface Props {
    launchData?: any;
}

interface launch {
    label: string;
    data: any[];
    backgroundColor: string;
}

interface locations {
    labels: any[];
    datasets: launch[]
}

const GraphView: React.FunctionComponent<Props> = (props) => {
    const launches = launchReducer(props.launchData);
    const canvas: React.Ref<HTMLCanvasElement> = React.createRef();

    const launchCount = (launches: any) => {
        let locations = []
        for (let launchLocation in launches) {
            locations.push(
                <dd key={launchLocation} className={`percentage percentage-${launches[launchLocation].length}`}>{launchLocation} {launches[launchLocation].length}</dd>
            )
        }
        return locations;
    }

    const launchesByLocation = (launches: any) => {
        let locations: locations = {
            labels: [],
            datasets: [
                {
                    label: 'launches',
                    data: [],
                    backgroundColor: '#FFF'
                }
            ],
        };
        for (let launchLocation in launches) {
            locations.labels.push(launchLocation);
            locations.datasets[0].data.push(launches[launchLocation].length);
        }
        return locations;
    }

    React.useEffect(() => {
        new Chart(canvas.current.getContext('2d'), {
            type: 'bar',
            data: launchesByLocation(launches),
            options: {
                scales: {
                    yAxes: [{
                        ticks: {
                            beginAtZero: true,
                            stepSize: 1
                        }
                    }]
                }
            }
        });
    });

    return (
        <div>
            <dl>
                <dt>Launches:</dt>
                {launchCount(launches)}
            </dl>
            <canvas id="chart" ref={canvas}></canvas>
        </div>
    );
};

export default GraphView;