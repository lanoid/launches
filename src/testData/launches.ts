const launches = [
    {
        name: 'Daft Punk',
        location: {
            name: 'My house'
        },
        windowstart: 'November 14, 2017 18:28:00 UTC',
        windowend: 'November 14, 2017 18:53:00 UTC'
    },
    {
        name: 'Daft Punk',
        location: {
            name: 'My house'
        },
        windowstart: 'November 14, 2017 18:28:00 UTC',
        windowend: 'November 14, 2017 18:53:00 UTC'
    },
    {
        name: 'Daft Punk',
        location: {
            name: 'Your house'
        },
        windowstart: 'November 14, 2017 18:28:00 UTC',
        windowend: 'November 14, 2017 18:53:00 UTC'
    }
];

export default launches;