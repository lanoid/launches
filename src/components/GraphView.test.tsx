import * as React from 'react';
import { shallow } from 'enzyme';
import GraphView from './GraphView';
import * as launches from '../testData/launches';

describe('<Graphview />', () => {
    const graphView = shallow(<GraphView launchData={launches.default} />);
    
    it('matches the snapshot', () => {
        expect(graphView).toMatchSnapshot();
    });
});