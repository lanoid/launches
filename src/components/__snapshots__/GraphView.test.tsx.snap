// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Graphview /> matches the snapshot 1`] = `
<div>
  <dl>
    <dt>
      Launches:
    </dt>
    <dd
      className="percentage percentage-2"
      key="My house"
    >
      My house
       
      2
    </dd>
    <dd
      className="percentage percentage-1"
      key="Your house"
    >
      Your house
       
      1
    </dd>
  </dl>
  <canvas
    id="chart"
  />
</div>
`;
